File: Cocoa/Plug-Ins/Template Editor/ElementBBIT.m
#import "ElementBBIT.h"
#import "ElementBFLG.h"

#define SIZE_ON_DISK (1)

@implementation ElementBBIT

+ (unsigned int)length {
    return 8;
}

- (instancetype)initForType:(NSString *)t withLabel:(NSString *)l
{
    if (self = [super initForType:t withLabel:l]) {
        self.first = YES;
        if ([[t substringFromIndex:1] isEqualToString:@"BIT"]) {
            self.bits = 1;
        } else {
            // XXnn - bit field or fill bits
            self.bits = [[t substringFromIndex:2] intValue];
            if (self.bits > self.class.length) return nil; // Too big
            self.visible = [t characterAtIndex:1] != 'F'; // Hide fill bits
        }
    }
    return self;
}

- (void)configureView:(NSView *)view
{
    if (_bits == self.class.length) {
        // Display as checkboxes
        NSRect frame = NSMakeRect(0, self.rowHeight-1, 20, 20);
        for (unsigned int i = 0; i < _bits; i++) {
            if (i % 8 == 0) {
                frame.origin.x = 0;
                frame.origin.y -= 20;
            }
            [view addSubview:[ElementBFLG createCheckboxWithFrame:frame forElement:self.bitList[i]]];
            frame.origin.x += 20;
        }
    } else if (_bits == 1) {
        [view addSubview:[ElementBFLG createCheckboxWithFrame:view.frame forElement:self]];
    } else {
        [super configureView:view];
        [view.subviews[0] setPlaceholderString:[NSString stringWithFormat:@"%d bits", _bits]];
    }
}

- (void)configure
{
    if (self.bits > 1 && self.bits < self.class.length)
        [super configure]; // Allow bitfields to configure cases
    if (!self.first) return;
    self.bitList = [NSMutableArray new];
    ElementBBIT *element;
    // Special treatment for BB08/WB16/LB32 (otherwise equivalent to UBYT/UWRD/ULNG): display as rows of checkboxes
    if (self.visible && _bits == self.class.length) {
        for (unsigned int pos = self.class.length; pos > 0; pos--) {
            element = [ElementBBIT elementForType:@"BBIT" withLabel:nil];
            element.position = pos - 1;
            [self.bitList addObject:element];
        }
        self.rowHeight = (_bits/8 * 20) + 2;
    } else {
        [self.bitList addObject:self];
        unsigned int pos = _position = self.class.length - _bits;
        while (pos > 0) {
            element = (ElementBBIT *)[self.parentList peek:self.bitList.count];
            if (element.class != self.class) {
                NSLog(@"Not enough bits in bit field.");
                break;
            }
            if (element.bits > pos) {
                NSLog(@"'%@' element creates too many bits to complete bit field.", element.type);
                break;
            }
            element.position = pos -= element.bits;
            element.first = NO;
            [self.bitList addObject:element];
        }
    }
}

- (void)readDataFrom:(ResourceStream *)stream
{
    if (!self.bitList) return;
    UInt8 completeValue = 0;
    [stream readAmount:SIZE_ON_DISK toBuffer:&completeValue];
    for (ElementBBIT* element in self.bitList) {
        element.value = (completeValue >> element.position) & ((1 << element.bits) - 1);
    }
}

- (void)sizeOnDisk:(UInt32 *)size
{
    if (!self.bitList) return;
    *size += SIZE_ON_DISK;
}

- (void)writeDataTo:(ResourceStream *)stream
{
    if (!self.bitList) return;
    UInt8 completeValue = 0;
    for (ElementBBIT* element in self.bitList) {
        completeValue |= element.value << element.position;
    }
    [stream writeAmount:SIZE_ON_DISK fromBuffer:&completeValue];
}

- (NSFormatter *)formatter
{
    NSNumberFormatter *formatter = [[NSNumberFormatter alloc] init];
    formatter.hasThousandSeparators = NO;
    formatter.minimum = 0;
    formatter.maximum = @((1 << _bits) - 1);
    formatter.nilSymbol = @"\0";
    return formatter;
}

@end

File: Plug-Ins/Template Editor/Elements/ElementBBIT.swift
import Cocoa
import RFSupport

// Implements BBIT/BBnn/BFnn, WBIT/WBnn/WFnn, LBIT/LBnn/LFnn
class ElementBBIT<T: FixedWidthInteger & UnsignedInteger>: RangeableElement {
    @objc private var value: UInt = 0
    private var bits = 1
    private var position = 0
    private var bitList: [ElementBBIT] = []
    private var first = true
    
    required init!(type: String, label: String, tooltip: String? = nil) {
        if !type.hasSuffix("BIT") {
            // XXnn - bit field or fill bits
            bits = Int(type.suffix(2))!
            guard bits <= T.bitWidth else {
                return nil
            }
        }
        super.init(type: type, label: label, tooltip: tooltip)
        self.visible = type.dropFirst().first != "F"
    }
    
    override func configure() throws {
        if bits == 1 {
            // Single bit, configure like BOOL
            try ElementBOOL.readRadioCases(for: self)
        } else if bits < T.bitWidth {
            // Allow bitfields to configure cases
            try super.configure()
        }
        if !first {
            return
        }
        // Special treatment for BB08/WB16/LB32 (otherwise equivalent to UBYT/UWRD/ULNG): display as rows of checkboxes
        if bits == T.bitWidth {
            for pos in (0..<T.bitWidth).reversed() {
                let bbit = Self.init(type: "BBIT", label: "")!
                bbit.position = pos
                bitList.append(bbit)
            }
            self.rowHeight = Double(bits/8 * 20) + 2
        } else {
            bitList.append(self)
            position = T.bitWidth - bits
            var pos = position
            var i = 0
            while pos > 0 {
                i += 1
                let next = self.parentList.peek(i)
                // Skip over cosmetic items
                if let next = next, ["CASE", "CASR", "DVDR", "RREF", "PACK"].contains(next.type) {
                    continue
                }
                guard let bbit = next as? ElementBBIT else {
                    throw TemplateError.invalidStructure(self, NSLocalizedString("Not enough bits in bit field.", comment: ""))
                }
                if bbit.bits > pos {
                    throw TemplateError.invalidStructure(bbit, NSLocalizedString("Too many bits in bit field.", comment: ""))
                }
                pos -= bbit.bits
                bbit.position = pos
                bbit.first = false
                bitList.append(bbit)
            }
        }
    }
    
    override func configure(view: NSView) {
        if bits == 1 {
            ElementBOOL.configureRadios(in: view, for: self)
        } else if bits == T.bitWidth {
            // Display as checkboxes
            var frame = view.frame
            frame.origin.y += 1
            frame.size.width = 20
            frame.size.height = 20
            for i in 0..<bits {
                view.addSubview(ElementBOOL.createCheckbox(with: frame, for: bitList[i]))
                frame.origin.x += 20
                if i % 8 == 7 {
                    frame.origin.x = view.frame.origin.x
                    frame.origin.y += 20
                }
            }
        } else {
            super.configure(view: view)
            if let field = view.subviews.first as? NSTextField {
                field.placeholderString = "\(bits) bits"
            }
        }
    }
    
    override func readData(from reader: BinaryDataReader) throws {
        if first {
            let completeValue: T = try reader.read()
            for bbit in bitList {
                bbit.value = UInt((completeValue >> bbit.position) & ((1 << bbit.bits) - 1))
            }
        }
    }
    
    override func writeData(to writer: BinaryDataWriter) {
        if first {
            var completeValue: T = 0
            for bbit in bitList {
                completeValue |= T(bbit.value << bbit.position)
            }
            writer.write(completeValue)
        }
    }
    
    override var formatter: Formatter? {
        if Element.sharedFormatters[type] == nil {
            let formatter = NumberFormatter()
            formatter.hasThousandSeparators = true
            formatter.minimum = 0
            formatter.maximum = (1 << bits) - 1 as NSNumber
            formatter.nilSymbol = "\0"
            Element.sharedFormatters[type] = formatter
        }
        return Element.sharedFormatters[type]
    }
}
